### PR TITLE
Jinghan/add a cache for offline stream features

### LINKED ISSUE #807: Add buffer to cache feature values for offline store

### PR DESCRIPTION
**Pull request by jinghancc**
This PR implements a cache for offline stream feature values, which inserts stream feature values to offline store in batch.

close #807

### CODE REVIEW

File: pkg/oomstore/cache.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+	go func() {
+		for {
+			select {
+			case <-cache.quit:
+				return
+			case <-cache.ticker.C:
+				for groupID, b := range cache.buffer {
+					if len(b.records) > 0 && time.Since(b.lastPush) > cache.minPeriod {
+						if err := cache.send(ctx, s, groupID); err != nil {
+							fmt.Printf("%+v", err)
```
Comment (wfxr) [inline pkg/oomstore/cache.go:63]: The following line can be used to print error to `stderr` before we have logging framework:
```go
fmt.Fprintf(os.Stderr, "Error pushing to offline store: %+v", err)
```

**Review by wfxr — COMMENTED** [review]
Forget to process table not exists error?
https://github.com/oom-ai/oomstore/blob/07f8eadda701a55d8ce6ad3db0876a676b8b71b6/internal/database/offline/sqlutil/push.go#L25

File: pkg/oomstore/cache.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+type Cache struct {
+	capacity  int
+	minPeriod time.Duration
+	mu        sync.Mutex
+	quit      chan bool
+
+	ch     chan types.StreamRecord // buffer channel
+	buffer map[int]*BufferGroup    // group_id -> buffered rows
+	ticker *time.Ticker            // group_id -> ticker
```
Comment (wfxr) [inline pkg/oomstore/cache.go:38]: Outdated comment

File: pkg/oomstore/cache.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+				for groupID, b := range cache.buffer {
+					if len(b.records) > 0 && time.Since(b.lastPush) > cache.minPeriod {
+						if err := cache.send(ctx, s, groupID); err != nil {
+							fmt.Printf("%+v", err)
+						}
+					}
+				}
+			case record := <-cache.ch:
+				groupID := record.GroupID
+				if len(cache.buffer[groupID].records) == cache.capacity {
```
Comment (wfxr) [inline pkg/oomstore/cache.go:69]: `len(cache.buffer[groupID].records) >= cache.capacity` should be better ?

File: pkg/oomstore/cache.go (hunk 2 above)
Comment (jinghancc) [reply pkg/oomstore/cache.go]: fixed

File: pkg/oomstore/cache.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+		buffer: make(map[int]*BufferGroup),
+		ticker: time.NewTicker(Period),
+	}
+	s.cache = cache
+
+	go func() {
+		for {
+			select {
+			case <-cache.quit:
+				return
```
Comment (lianxmfor) [inline pkg/oomstore/cache.go:58]: Before exiting, need to check if there are any records that have not been sent.

**Comment by jinghancc** [comment]
> Forget to process table not exists error?
> 
> https://github.com/oom-ai/oomstore/blob/07f8eadda701a55d8ce6ad3db0876a676b8b71b6/internal/database/offline/sqlutil/push.go#L25

@wfxr `newRevision` creates new offline cdc table and revision

File: pkg/oomstore/cache.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+		}
+	}
+
+	c.buffer[groupID].records = make([]types.StreamRecord, 0, c.capacity)
+	c.buffer[groupID].lastPush = time.Now()
+	return err
+}
+
+func (c *Cache) newRevision(ctx context.Context, s *OomStore, groupID int, revision int64) error {
+	features, err := s.metadata.ListFeature(ctx, metadata.ListFeatureOpt{
```
Comment (YikSanChan) [inline pkg/oomstore/cache.go:143]: `features` can be passed down as arguments

File: pkg/oomstore/cache.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+	}()
+}
+
+func (c *Cache) Close() error {
+	c.ticker.Stop()
+	c.quit <- true
+	return nil
+}
+
+func (c *Cache) Push(groupID int, record types.StreamRecord) {
```
Comment (lianxmfor) [inline pkg/oomstore/cache.go:86]: `types.StreamRecord` already has the field `GroupID`, so you can remove the `groupID` parameter?

**Comment by wfxr** [comment]
> > Forget to process table not exists error?
> > https://github.com/oom-ai/oomstore/blob/07f8eadda701a55d8ce6ad3db0876a676b8b71b6/internal/database/offline/sqlutil/push.go#L25
> 
> @wfxr `newRevision` creates new offline cdc table and revision

@jinghancc Got it.

**Review by YikSanChan — APPROVED** [review]
lgtm

File: pkg/oomstore/cache.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+)
+
+type BufferGroup struct {
+	groupID  int
+	records  []types.StreamRecord
+	lastPush time.Time
+	mu       sync.Mutex
+}
+
+type Cache struct {
```
Comment (wfxr) [inline pkg/oomstore/cache.go:30]: Naming suggestion: `StreamPushProcessor`

File: pkg/oomstore/cache.go [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+						fmt.Printf("%+v", err)
+					}
+				}
+				cache.buffer[groupID].records = append(cache.buffer[groupID].records, record)
+			}
+		}
+	}()
+}
+
+func (c *Cache) Close() error {
```
[the remark this answers is not in the document]
Comment (jinghancc) [reply pkg/oomstore/cache.go:80]: to return `error` so that it implements `io.Closer`

File: pkg/oomstore/cache.go (hunk 7 above)
Comment (jinghancc) [reply pkg/oomstore/cache.go]: updated

File: pkg/oomstore/cache.go (hunk 6 above)
Comment (jinghancc) [reply pkg/oomstore/cache.go]: fixed

File: pkg/oomstore/cache.go (hunk 4 above)
Comment (jinghancc) [reply pkg/oomstore/cache.go]: fixed

File: pkg/oomstore/cache.go (hunk 3 above)
Comment (jinghancc) [reply pkg/oomstore/cache.go]: updated

**Review by lianxmfor — APPROVED** [review]
lgtm

**Review by wfxr — APPROVED** [review]
LGTM

**Pull request merged**